File: eb-application/src/main/java/com/ebnbin/ebapplication/view/StateView.java
package com.ebnbin.ebapplication.view;

import android.content.Context;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.annotation.StyleRes;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.ebnbin.ebapplication.R;

/**
 * A FrameLayout that switches states easily.
 */
public final class StateView extends FrameLayout {
    private final LayoutInflater mLayoutInflater;

    public StateView(@NonNull Context context) {
        super(context);

        mLayoutInflater = LayoutInflater.from(context);
    }

    public StateView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        mLayoutInflater = LayoutInflater.from(context);
    }

    public StateView(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mLayoutInflater = LayoutInflater.from(context);
    }

    public StateView(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr,
            @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        mLayoutInflater = LayoutInflater.from(context);
    }

    //*****************************************************************************************************************
    // States.

    private enum State {
        /**
         * No state.
         */
        NONE,
        /**
         * Loading state.
         */
        LOADING,
        /**
         * Failure state.
         */
        FAILURE,
        /**
         * Progressing state.
         */
        PROGRESSING,
        /**
         * No data state.
         */
        NO_DATA
    }

    @NonNull
    private State mState = State.NONE;

    //*****************************************************************************************************************

    /**
     * Saves visibilities of child views for restoring if they are set {@link #GONE}.
     */
    private int[] mVisibilities;

    /**
     * Saves visibilities of child views and sets them {@link #GONE}.
     */
    private void hideChildViews() {
        if (mVisibilities != null) {
            return;
        }

        int count = getChildCount();
        mVisibilities = new int[count];

        for (int i = 0; i < count; i++) {
            View childView = getChildAt(i);

            mVisibilities[i] = childView.getVisibility();
            childView.setVisibility(GONE);
        }
    }

    /**
     * Restores visibilities of child views.
     */
    private void showChildViews() {
        if (mVisibilities == null || mVisibilities.length != getChildCount()) {
            return;
        }

        for (int i = 0; i < mVisibilities.length; i++) {
            getChildAt(i).setVisibility(mVisibilities[i]);
        }

        mVisibilities = null;
    }

    //*****************************************************************************************************************
    // Switch mode.

    private enum SwitchMode {
        /**
         * Keeps visibilities of child views.
         */
        KEEP,
        /**
         * Shows child views.
         */
        OVERLAY,
        /**
         * Hides child views.
         */
        REPLACE
    }

    private void setSwitchMode(@NonNull SwitchMode switchMode) {
        switch (switchMode) {
            case OVERLAY: {
                showChildViews();

                break;
            }
            case REPLACE: {
                hideChildViews();

                break;
            }
            case KEEP:
            default: {
                break;
            }
        }
    }

    //*****************************************************************************************************************
    // Clears state and shows child views.

    public void stateNone() {
        if (mState == State.NONE) {
            return;
        }

        removeViewAt(getChildCount() - 1);

        showChildViews();

        mState = State.NONE;
    }

    //*****************************************************************************************************************
    // Loading state.

    private FrameLayout mLoadingFrameLayout;

    public void stateLoading() {
        if (!switchState(State.LOADING, SwitchMode.REPLACE)) {
            return;
        }

        if (mLoadingFrameLayout == null) {
            mLoadingFrameLayout = (FrameLayout) mLayoutInflater.inflate(R.layout.eb_view_state_loading, this, false);
        }

        addView(mLoadingFrameLayout);
    }

    //*****************************************************************************************************************
    // Failure state.

    private FrameLayout mFailureFrameLayout;
    private ImageView mRefreshImageView;
    private TextView mFailureTextView;

    public void stateFailure(@Nullable OnClickListener onRefreshClickListener) {
        stateFailure(onRefreshClickListener, null);
    }

    public void stateFailure(@Nullable OnClickListener onRefreshClickListener, @StringRes int failureStringId) {
        stateFailure(onRefreshClickListener, failureStringId == -1 ? null
                : failureStringId == 0 ? "" : getContext().getString(failureStringId));
    }

    public void stateFailure(@Nullable OnClickListener onRefreshClickListener, @Nullable String failureString) {
        if (!switchState(State.FAILURE, SwitchMode.REPLACE)) {
            return;
        }

        if (mFailureFrameLayout == null) {
            mFailureFrameLayout = (FrameLayout) mLayoutInflater.inflate(R.layout.eb_view_state_failure, this, false);
            mRefreshImageView = (ImageView) mFailureFrameLayout.findViewById(R.id.eb_refresh);
            mFailureTextView = mFailureFrameLayout.findViewById(R.id.eb_failure);
        }
        mRefreshImageView.setOnClickListener(onRefreshClickListener);
        if (failureString != null) {
            mFailureTextView.setText(failureString);
        }

        addView(mFailureFrameLayout);
    }

    //*****************************************************************************************************************
    // No data state.

    private FrameLayout mNoDataFrameLayout;
    private TextView mNoDataTextView;

    public void stateNoData() {
        stateNoData(null);
    }

    public void stateNoData(@StringRes int noDataStringId) {
        stateNoData(noDataStringId == -1 ? null : noDataStringId == 0 ? "" : getContext().getString(noDataStringId));
    }

    public void stateNoData(@Nullable String noDataString) {
        if (!switchState(State.NO_DATA, SwitchMode.KEEP)) {
            return;
        }

        if (mNoDataFrameLayout == null) {
            mNoDataFrameLayout = (FrameLayout) mLayoutInflater.inflate(R.layout.eb_view_state_no_data, this, false);
            mNoDataTextView = mNoDataFrameLayout.findViewById(R.id.eb_no_data);
        }
        if (noDataString != null) {
            mNoDataTextView.setText(noDataString);
        }

        addView(mNoDataFrameLayout);
    }

    //*****************************************************************************************************************
    // Progressing state.

    private FrameLayout mProgressingFrameLayout;
    private ProgressBar mProgressBar;

    public void stateProgressing() {
        if (!switchState(State.PROGRESSING, SwitchMode.OVERLAY)) {
            return;
        }

        if (mProgressingFrameLayout == null) {
            mProgressingFrameLayout = (FrameLayout) mLayoutInflater.inflate(R.layout.eb_view_state_progressing, this,
                    false);
            mProgressBar = (ProgressBar) mProgressingFrameLayout.findViewById(R.id.progress_bar);
        }

        addView(mProgressingFrameLayout);
    }

    public void stateProgressing(int progress) {
        if (mState != State.PROGRESSING) {
            return;
        }

        if (progress == -1) {
            mProgressBar.setIndeterminate(true);
        } else {
            mProgressBar.setIndeterminate(false);
            mProgressBar.setProgress(progress);
        }
    }

    //*****************************************************************************************************************
    // Not none state.

    /**
     * General method for all states except {@link State#NONE}.
     *
     * @return Whether state has changed.
     */
    private boolean switchState(@NonNull State state, @NonNull SwitchMode switchMode) {
        if (mState == state) {
            return false;
        }

        if (mState != State.NONE) {
            removeViewAt(getChildCount() - 1);
        }

        mState = state;

        setSwitchMode(switchMode);

        return true;
    }
}

File: eb-application/src/main/java/com/ebnbin/ebapplication/view/StateView.kt
package com.ebnbin.ebapplication.view

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import com.ebnbin.eb.util.EBRuntimeException
import com.ebnbin.ebapplication.R

/**
 * An extended [FrameLayout] that switches states easily.
 */
open class StateView : FrameLayout {
    private val layoutInflater: LayoutInflater

    constructor(context: Context?) : this(context, null)

    constructor(context: Context?, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : this(context, attrs, defStyleAttr, 0)

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes) {
        layoutInflater = LayoutInflater.from(context)
    }

    //*****************************************************************************************************************
    // States.

    enum class State {
        /**
         * No state.
         */
        NONE,
        /**
         * Shows a [ProgressBar] and an optional hint. Child views will be hidden.
         */
        LOADING,
        /**
         * Shows a horizontal style [ProgressBar] at view top and shows child views.
         */
        PROGRESSING,
        /**
         * Shows a refresh [View] and an optional hint. Child views will be hidden.
         */
        FAILURE,
        /**
         * Shows a hint for no data states and hides child views.
         */
        NO_DATA
    }

    /**
     * Each state should have and only have a layer [ViewGroup] added to current [FrameLayout] except [State.NONE].
     */
    var state = State.NONE
        private set

    /**
     * Returns whether there is a state set currently.
     */
    fun hasState(): Boolean {
        return state != State.NONE
    }

    /**
     * Clears current state.
     */
    fun clearState() {
        if (state == State.NONE) {
            return
        }

        removeViewAt(childCount - 1)

        state = State.NONE

        setChildViewsHidden(false)
    }

    /**
     * General method for all states except [State.NONE].
     *
     * @param state Should NOT be [State.NONE].
     * @param hideChildViews Whether to hide child views.
     */
    private fun switchState(state: State, hideChildViews: Boolean) {
        if (state == State.NONE) {
            throw EBRuntimeException()
        }

        clearState()

        setChildViewsHidden(hideChildViews)

        this.state = state
    }

    //*****************************************************************************************************************

    /**
     * Saved visibilities of child views.
     */
    private var childViewsVisibilities: MutableList<Int>? = null

    /**
     * Returns whether child views has been hidden.
     */
    fun isChildViewsHidden(): Boolean {
        return childViewsVisibilities != null
    }

    /**
     * Sets visibilities of child views. Only works if there is no state currently.
     *
     * @param hidden If `true`, saves visibilities of child views and sets them [View.GONE]. Otherwise, restores
     * visibilities of child views that saved. If visibilities has been saved, NEITHER change visibilities of child
     * views NOR add or remove child views.
     */
    fun setChildViewsHidden(hidden: Boolean) {
        if (hasState()) {
            return
        }

        if (hidden) {
            if (childViewsVisibilities != null) {
                return
            }

            childViewsVisibilities = ArrayList<Int>(childCount)

            for (i in 0..childCount - 1) {
                val childView = getChildAt(i)

                childViewsVisibilities!!.add(childView.visibility)
                childView.visibility = View.GONE
            }
        } else {
            if (childViewsVisibilities == null) {
                return
            }

            if (childViewsVisibilities!!.size != childCount) {
                throw EBRuntimeException()
            }

            for (i in childViewsVisibilities!!.indices) {
                getChildAt(i).visibility = childViewsVisibilities!![i]
            }

            childViewsVisibilities = null
        }
    }

    //*****************************************************************************************************************
    // State LOADING.

    private var loadingViewGroup: ViewGroup? = null
    private var loadingHintTextView: TextView? = null

    /**
     * Default hint for state [State.LOADING].
     */
    var defLoadingHint: String = context.getString(R.string.eb_state_view_hint_loading)

    /**
     * Switches to state [State.LOADING].
     *
     * @param hint Optional hint. If `null`, default value will be used. If not set, `null` will be used.
     */
    fun stateLoading(hint: String? = null) {
        switchState(State.LOADING, true)

        if (loadingViewGroup == null) {
            loadingViewGroup = layoutInflater.inflate(R.layout.eb_state_view_loading, this, false) as ViewGroup
            loadingHintTextView = loadingViewGroup!!.findViewById(R.id.eb_hint) as TextView
        }

        val tmpHint = hint ?: defLoadingHint
        if (tmpHint == EMPTY_HINT) {
            loadingHintTextView!!.visibility = View.GONE
            loadingHintTextView!!.text = null
        } else {
            loadingHintTextView!!.visibility = View.VISIBLE
            loadingHintTextView!!.text = tmpHint
        }

        addView(loadingViewGroup)
    }

    //*****************************************************************************************************************
    // State PROGRESSING.

    private var progressingViewGroup: ViewGroup? = null
    private var progressingProgressBar: ProgressBar? = null

    /**
     * Switches to state [State.PROGRESSING].
     *
     * @param progress If `-1`, [ProgressBar] will be indeterminate mode. Otherwise, be between `0` and `100` to
     * disable indeterminate mode. If not set, `-1` will be used.
     * @param updateProgressOnly If `true`, only to change progress if current state is [State.PROGRESSING] already.
     * Otherwise, switches to state [State.PROGRESSING] firstly. If not set, `false` will be used.
     */
    fun stateProgressing(progress: Int = -1, updateProgressOnly: Boolean = false) {
        if (updateProgressOnly) {
            if (state != State.PROGRESSING) {
                return
            }
        } else {
            switchState(State.PROGRESSING, false)

            if (progressingViewGroup == null) {
                progressingViewGroup = layoutInflater.inflate(R.layout.eb_state_view_progressing, this, false)
                        as ViewGroup
                progressingProgressBar = progressingViewGroup!!.findViewById(R.id.progress_bar) as ProgressBar
            }

            addView(progressingViewGroup)
        }

        if (progress == -1) {
            progressingProgressBar!!.isIndeterminate = true
            progressingProgressBar!!.progress = 0
        } else {
            progressingProgressBar!!.isIndeterminate = false
            progressingProgressBar!!.progress = progress
        }
    }

    //*****************************************************************************************************************
    // State FAILURE.

    /**
     * Default [OnRefreshListener] for state [State.FAILURE].
     */
    var defOnRefreshListener: OnRefreshListener = EMPTY_ON_REFRESH_LISTENER
    /**
     * Default hint for  state [State.FAILURE].
     */
    var defFailureHint: String = context.getString(R.string.eb_state_view_hint_failure)

    /**
     * Switches to state [State.FAILURE].
     *
     * @param onRefreshListener [View.OnClickListener] of the refresh [View]. If `null`, default value will be used. If
     * not set, `null` will be used.
     * @param hint Optional hint. If `null`, default value will be used. If not set, `null` will be used.
     */
    fun stateFailure(onRefreshListener: OnRefreshListener? = null, hint: String? = null) {
        val tmpOnRefreshListener = onRefreshListener ?: defOnRefreshListener
        val tmpHint = hint ?: defFailureHint
        stateMessage(State.FAILURE, tmpOnRefreshListener, tmpHint)
    }

    /**
     * On refresh listener for state [State.FAILURE].
     */
    interface OnRefreshListener {
        fun onRefresh()
    }

    //*****************************************************************************************************************
    // State NO_DATA.

    /**
     * Default hint for  state [State.NO_DATA].
     */
    var defNoDataHint: String = context.getString(R.string.eb_state_view_hint_no_data)

    /**
     * Switches to state [State.NO_DATA].
     *
     * @param hint Hint. If `null`, default value will be used. If not set, `null` will be used.
     */
    fun stateNoData(hint: String? = null) {
        val tmpHint = hint ?: defNoDataHint
        stateMessage(State.NO_DATA, EMPTY_ON_REFRESH_LISTENER, tmpHint)
    }

    //*****************************************************************************************************************
    // State FAILURE or NO_DATA.

    private var messageViewGroup: ViewGroup? = null
    private var messageRefreshView: View? = null
    private var messageHintTextView: TextView? = null

    /**
     * Switches to state [State.FAILURE] or [State.NO_DATA].
     *
     * @param state Should be [State.FAILURE] or [State.NO_DATA].
     * @param onRefreshListener If non-empty, shows a refresh [View] and provides its [View.OnClickListener].
     * @param hint Optional hint.
     */
    private fun stateMessage(state: State, onRefreshListener: OnRefreshListener, hint: String) {
        if (state != State.FAILURE && state != State.NO_DATA) {
            throw EBRuntimeException()
        }

        switchState(state, true)

        if (messageViewGroup == null) {
            messageViewGroup = layoutInflater.inflate(R.layout.eb_state_view_message, this, false) as ViewGroup
            messageRefreshView = messageViewGroup!!.findViewById(R.id.eb_refresh)
            messageHintTextView = messageViewGroup!!.findViewById(R.id.eb_hint) as TextView
        }

        if (onRefreshListener == EMPTY_ON_REFRESH_LISTENER) {
            messageRefreshView!!.visibility = View.GONE
            messageRefreshView!!.setOnClickListener(null)
        } else{
            messageRefreshView!!.visibility = View.VISIBLE
            messageRefreshView!!.setOnClickListener { onRefreshListener.onRefresh() }
        }

        if (hint == EMPTY_HINT) {
            messageHintTextView!!.visibility = View.GONE
            messageHintTextView!!.text = null
        } else {
            messageHintTextView!!.visibility = View.VISIBLE
            messageHintTextView!!.text = hint
        }

        addView(messageViewGroup)
    }

    //*****************************************************************************************************************

    companion object {
        //*************************************************************************************************************
        // Empty value for states.

        const val EMPTY_HINT = ""
        val EMPTY_ON_REFRESH_LISTENER = object : StateView.OnRefreshListener {
            override fun onRefresh() {}
        }
    }
}
